### PR TITLE
Miscellaneous

### PR DESCRIPTION
**Pull request by davidwalter2**
Bugfix
- Turn off region swapping in ABCD method in case smoothing mode is different to fakerate or hybrid smoothing; This fixes a bug in the binned ABCD methods

Functionality
- Exclude prompt muons form QCD MC in w histmaker by default

Plotting
- Style improvements on the impact pulls and constraints plots
- Fixing legend padding; now by default the "empty" places in the legend are towards the center of the figure

Code style
- Move parser functions into seperate file utilities/parsing.py to factorize the code better

Pre commit hooks
- Run flake8 only on staged files; Currently it's run on all files but local files may not be desired to fulfill all requirements.
- Additional checks in flake8, such as check for undefined variables (including fixes for these instances)
- use pylint to give a rating of the staged code

### CODE REVIEW

**Comment by davidwalter2** [comment]
The PR validation is here for this PR:
https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/PR555/2024_10_30/
compared to the reference:
https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/ReferenceRuns/2024_10_30_1ffad3b/

While the high pileup results are the same (only style of impact plots change) there is a big effect in the low pileup, improving the data-to-MC agreement. This is due to the fix of the binned ABCD method (I'm sorry @jeyserma for introducing this bug ...). 

From my side it can be merged

File: utilities/styles/nuisance_translate.json [hunk 1]
```diff
@@ -56,7 +56,118 @@
...
+    "QCDscaleWinclusive_PtV0_13000helicity_0_SymAvg": "<i>A</i><sub>0</sub> W angular coeff., inc.",
+    "QCDscaleWinclusive_PtV0_13000helicity_2_SymAvg": "<i>A</i><sub>2</sub> W angular coeff., inc.",
+    "scetlibNPgamma": "SCETLib γ",
+    "chargeVgenNP0scetlibNPZLambda2": "SCETLib λ²(Z)",
+    "chargeVgenNP1scetlibNPWLambda2": "SCETLib λ²(W⁻)",
+    "chargeVgenNP0scetlibNPWLambda2": "SCETLib λ²(W⁺)",
+    "chargeVgenNP0scetlibNPWDelta_Lambda2": "SCETLib Δλ²(W⁻)",
+    "chargeVgenNP1scetlibNPWDelta_Lambda2": "SCETLib Δλ²(W⁺)",
+    "chargeVgenNP0scetlibNPZDelta_Lambda2": "SCETLib Δλ²(Z)",
+    "chargeVgenNP0scetlibNPWLambda4": "SCETLib λ⁴(W⁻)",
```
[the remark this answers is not in the document]
Comment (davidwalter2) [reply utilities/styles/nuisance_translate.json:150]: Yes this works for plotly and avoids the need for MathJax which had different problems

**Comment by bendavid** [comment]
Did the blue and red colors swap for the impact plots?
(and is it now as desired, presumably blue should mean +1 sigma variation of the nuisance parameter has a positive impact on mW and red means it has a negative impact)

(I still think the legend is confusing/ambiguous, but we can deal with that later of course)

Even if it takes up more space it might be clearer to just have a two sided plot for the impacts part of the plot...

**Comment by bendavid** [comment]
There's another practical issue here.

I at least have been relying on the numerical values printed on the impact plots to see if there were really zero changes numerically.  I guess the info is still available from the logs, but this makes it more difficult/time consuming to check.  Maybe it can be a command line option in the script to still keep the numerical values on the plot (and turned on by default and/or specified on the command line options for the ci)

**Comment by davidwalter2** [comment]
> Did the blue and red colors swap for the impact plots? (and is it now as desired, presumably blue should mean +1 sigma variation of the nuisance parameter has a positive impact on mW and red means it has a negative impact)
> 
> (I still think the legend is confusing/ambiguous, but we can deal with that later of course)
> 
> Even if it takes up more space it might be clearer to just have a two sided plot for the impacts part of the plot...
Yes the colors are swapped, I think is is correct with the legend.
The way to read the plot and legend is that only positive variations of mW are shown, the effect of a +1 sigma variation of the nuisance parameter is shown in blue and of a -1 sigma variation of the nuisance parameter in red. 
The two sided impacts are still supported and are even the default (when removing the --oneSidedImpacts argument).

**Comment by bendavid** [comment]
Small point which can be addressed in the next PR:  The total impact for the wlike doesn't appear e.g. in https://cmsmwbot.web.cern.ch/WMassAnalysis/PRValidation/PR555/2024_10_31/impactsWlike_group_massShiftZ100MeV_noi.png

Maybe this is related to improperly taking the impact of the noi on itself as the total uncertainty rather than properly taking the error on the mass parameter?

**Pull request merged**